File: 프로그래머스/1/160586. 대충 만든 자판/대충 만든 자판.java
import java.util.*;

class Solution {
    public int[] solution(String[] keymap, String[] targets) {
        int[] answer = new int[targets.length];
        Map<Character, Integer> map = new HashMap<>();
        
        for(int i = 0; i < keymap.length; i++) {
            String str = keymap[i];
            
            for(int j = 0; j < str.length(); j++) {
                char ch = str.charAt(j);
                
                if(map.containsKey(ch)) map.put(ch, Math.min(j+1, map.get(ch)));
                else map.put(ch, j+1);
            }
        }
        
        for(int i = 0; i < targets.length; i++) {
            String str = targets[i];
            int sum = 0;
            
            for(int j = 0; j < str.length(); j++) {
                char ch = str.charAt(j);
                
                if(!map.containsKey(ch)) {
                    sum = -1;
                    break;
                }
                
                sum+=map.get(ch);
            }
            
            answer[i] = sum;
        }
        
        return answer;
    }
}

File: 프로그래머스/1/160586. 대충 만든 자판/대충 만든 자판.js
function solution(keymap, targets) {
    var answer = [];
    const map = new Map();
    
    for(let str of keymap) {
        for(let i = 0; i < str.length; i++) {
            const char = str[i];
            if(!map.has(char) || map.get(char) > i + 1) map.set(char, i+1);
        }
    }
    
    for(let target of targets) {
        let sum = 0;
        
        for(let char of target) {
            if(!map.has(char)) {
                sum = -1;
                break;
            }
            sum+=map.get(char);
        }
        
        answer.push(sum);
    }
    
    return answer;
}